base/exporter.go

```diff
@@ -0,0 +1,714 @@
+	ImportCount          *SgwIntStat `json:"import_count"`
+	ImportCancelCAS      *SgwIntStat `json:"import_cancel_cas"`
+	ImportErrorCount     *SgwIntStat `json:"import_error_count"`
+	ImportProcessingTime *SgwIntStat `json:"import_processing_time"`
+	ImportHighSeq        *SgwIntStat `json:"import_high_seq"`
+	ImportPartitions     *SgwIntStat `json:"import_partitions"`
+}
+
+type SgwStat struct {
+	statFQN       string
+	statDesc      *prometheus.Desc
+	dbName        string
+	statValueType prometheus.ValueType
+	mutex         sync.Mutex
+}
+
+type SgwIntStat struct {
+	SgwStat
+	Val int64
+}
+
+type SgwFloatStat struct {
+	SgwStat
+	Val float64
+}
+
+// Just a bool wrapper. Prometheus doesn't support boolean metrics and so this just goes to expvars
+type SgwBoolStat struct {
+	Val bool
+}
+
+func NewIntStat(subsystem string, key string, dbName string, statValueType prometheus.ValueType, initialValue int64) *SgwIntStat {
+	var label []string
+	if dbName != "" {
+		label = []string{"database"}
+	}
+
+	name := prometheus.BuildFQName("sgw", subsystem, key)
+	desc := prometheus.NewDesc(name, key, label, nil)
+
+	stat := &SgwIntStat{
+		SgwStat: SgwStat{
+			statFQN:       name,
+			statDesc:      desc,
+			dbName:        dbName,
+			statValueType: statValueType,
+		},
+		Val: initialValue,
+	}
+	prometheus.MustRegister(stat)
+	return stat
+}
+
+func (s *SgwIntStat) Describe(ch chan<- *prometheus.Desc) {
+	return
+}
+
+func (s *SgwIntStat) Collect(ch chan<- prometheus.Metric) {
+	var labelVals []string
+	if s.dbName != "" {
+		labelVals = append(labelVals, s.dbName)
+	}
+	ch <- prometheus.MustNewConstMetric(s.statDesc, s.statValueType, float64(s.Val), labelVals...)
+}
+
+func (s *SgwIntStat) Set(newV int64) {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+	s.Val = newV
+}
+
+func (s *SgwIntStat) SetIfMax(newV int64) {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+	if newV > s.Val {
+		s.Val = newV
+	}
+}
+
+func (s *SgwIntStat) Add(newV int64) {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+	s.Val += newV
+}
+
+func (s *SgwIntStat) MarshalJSON() ([]byte, error) {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+	return []byte(fmt.Sprintf("%v", s.Val)), nil
+}
+
+func (s *SgwIntStat) String() string {
+	return fmt.Sprintf("%v", s.Val)
+}
+
+func (s *SgwIntStat) Value() int64 {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+	return s.Val
+}
+
+func NewFloatStat(subsystem string, key string, dbName string, statValueType prometheus.ValueType, initialValue float64) *SgwFloatStat {
+	var label []string
+	if dbName != "" {
+		label = []string{"database"}
+	}
+
+	name := prometheus.BuildFQName("sgw", subsystem, key)
+	desc := prometheus.NewDesc(name, key, label, nil)
+
+	stat := &SgwFloatStat{
+		SgwStat: SgwStat{
+			statFQN:       name,
+			statDesc:      desc,
+			dbName:        dbName,
+			statValueType: statValueType,
+		},
+		Val: initialValue,
+	}
+	prometheus.MustRegister(stat)
+	return stat
+}
+
+func (s *SgwFloatStat) Describe(ch chan<- *prometheus.Desc) {
```

Comment:
There seems to be more in common than different between the SgwFloatStat and SgwIntStat implementations.  Can some or all of the common code be pushed up to SgwStat?

Reply:
Some of the constructor can be moved out